sympy/matrices/matrices.py

```diff
@@ -2044,6 +2192,425 @@ def minor_submatrix(self, i, j):
+        while piv_col < cols and piv_row < rows:
+            pivot_offset, pivot_val, \
+            assumed_nonzero, newly_determined = _find_reasonable_pivot(
+                get_col(piv_col)[piv_row:], iszerofunc, simpfunc)
+
+            # _find_reasonable_pivot may have simplified some things
+            # in the process.  Let's not let them go to waste
+            for (offset, val) in newly_determined:
+                offset += piv_row
+                mat[offset*cols + piv_col] = val
+
+            if pivot_offset is None:
+                piv_col += 1
+                continue
+
+            pivot_cols.append(piv_col)
+            if pivot_offset != 0:
+                row_swap(piv_row, pivot_offset + piv_row)
+                swaps.append((piv_row, pivot_offset + piv_row))
+
+            # if we aren't normalizing last, we normalize
+            # before we zero the other rows
+            if normalize_last is False:
+                i, j = piv_row, piv_col
+                mat[i*cols + j] = S.One
+                for p in range(i*cols + j + 1, (i + 1)*cols):
+                    mat[p] = mat[p] / pivot_val
+                # after normalizing, the pivot value is 1
+                pivot_val = S.One
+
+            # zero above and below the pivot
+            for row in range(rows):
+                # don't zero our current row
+                if row == piv_row:
+                    continue
+                # don't zero above the pivot unless we're told.
+                if zero_above is False and row < piv_row:
+                    continue
+                # if we're already a zero, don't do anything
+                val = mat[row*cols + piv_col]
+                if iszerofunc(val):
+                    continue
+
+                cross_cancel(pivot_val, row, val, piv_row)
+            piv_row += 1
+
+        # normalize each row
+        if normalize_last is True and normalize is True:
+            for piv_i, piv_j in enumerate(pivot_cols):
+                pivot_val = mat[piv_i*cols + piv_j]
+                mat[piv_i*cols + piv_j] = S.One
+                for p in range(piv_i*cols + piv_j + 1, (piv_i + 1)*cols):
+                    mat[p] = mat[p] / pivot_val
+
+        return self._new(self.rows, self.cols, mat), tuple(pivot_cols), tuple(swaps)
+
+    def echelon_form(self, iszerofunc=_iszero, simplify=False):
+        """Returns a matrix row-equivalent to `self` that is
+        in echelon form.  Note that echelon form of a matrix
+        is *not* unique, however, properties like the row
+        space and the null space are preserved."""
+        simpfunc = simplify if isinstance(
+            simplify, FunctionType) else _simplify
+
+        mat, pivots, swaps = self._eval_echelon_form(iszerofunc, simpfunc)
+        return mat
+
+    def elementary_col_op(self, op="n->kn", col=None, k=None, col1=None, col2=None):
+        """Perfoms the elementary row operation `op`.
```

Comment:
The name refers to columns. Should there be two methods, one for columns and another for rows?